### PR TITLE
[GPII-3697]: Use us-east1 for dev clusters, various fixes

### PR DESCRIPTION
**Pull request by natarajaya**
Summary of changes:
* Use single `TF_VAR_infra_region` variable to set region for different infra components.
* Drop deprecated `additional_zones` parameter from `gke-cluster` module.
* Use 3-node `n1-highcpu-8` clusters for stg and prd.
* Use 3-node `n1-highcpu-4` clusters for dev.
* Comment and README updates.

Deployment plan:
- [x] Schedule planned maintenance (`gke-cluster` module changes will cause cluster recreation).
- [x] Merge exekube PR.
- [ ] Disable cluster destruction protector.
- [ ] Merge this PR and let CI do its job.
- [ ] Enable cluster destruction protector.

There is a chance that switching region is not going to help with resource exhaustion issue too (#284), because initially I tried `us-west1` and got the same "Google Compute Engine does not have enough resources available" error.

### CODE REVIEW

File: gcp/live/dev/k8s/cluster/terraform.tfvars [hunk 1]
```diff
@@ -19,4 +19,7 @@ terragrunt = {
 
 # ↓ Module configuration (empty means all default)
 
-node_type = "n1-standard-2"
+node_type          = "n1-standard-2"
+initial_node_count = 2
```
Comment (mrtyler) [inline gcp/live/dev/k8s/cluster/terraform.tfvars]: I don't understand why this is 2 in dev, but 1 in stg and prd.

Comment (mrtyler) [reply gcp/live/dev/k8s/cluster/terraform.tfvars]: Is 2 enough nodes to have room for everything? I remember we had to add a fourth zone to facilitate rolling-updates and the like.

File: gcp/live/dev/infra/network/terraform.tfvars [hunk 2]
```diff
@@ -23,5 +23,8 @@ create_static_ip_address = true
 # DNS zone names are dynamic (e.g. mrtyler.dev.gcp.gpii.net) so they will be
 # injected from outside. (If they're not, defaulting to no DNS records is
 # reasonable.)
-dns_zones = {}
+dns_zones   = {}
 dns_records = {}
+
+static_ip_region = "us-east1"
+cluster_subnets  = { "0" = "us-east1,10.16.0.0/20,10.17.0.0/16,10.18.0.0/16" }
```
Comment (stepanstipl) [inline gcp/live/dev/infra/network/terraform.tfvars]: Since we use the region from `static_ip_region` with always the same set of subnets here (`10.16.0.0/20,10.17.0.0/16,10.18.0.0/16`), can we get rid of this and set it as default? Would save us repeating this multiple times.

File: shared/rakefiles/entrypoint.rake [hunk 3]
```diff
@@ -101,7 +101,10 @@ task :display_cluster_info do
   puts "Flowmanager endpoint:"
   puts "  curl -k https://flowmanager.#{ENV["TF_VAR_domain_name"] }"
   puts
-  puts "Run `rake destroy` to delete all the expensive resources created by the deployment"
+  puts "Run `rake test_preferences` to execute Locust tests for Preferences."
```
Comment (stepanstipl) [inline shared/rakefiles/entrypoint.rake]: Why is this included here? It seems to be totally unrelated change to changing regions.

File: gcp/live/dev/k8s/cluster/terraform.tfvars (hunk 1 above)
Comment (natarajaya) [reply gcp/live/dev/k8s/cluster/terraform.tfvars]: We need at least 4 `n1-standard-2` nodes to have room for everything.

There are 4 zones in us-central1 (`additional_zones   = ["us-central1-a", "us-central1-b", "us-central1-c", "us-central1-f"]`), with `initial_node_count = 1` it makes 4 worker nodes total.

But there are only 3 zones in us-east1 (b,c,d), so to have 4 worker nodes I changed initial node count to 2 and used only us-east1-b and us-east1-c as additional zones.

File: shared/rakefiles/entrypoint.rake (hunk 3 above)
Comment (natarajaya) [reply shared/rakefiles/entrypoint.rake]: It is unrelated. But it is just a minor comment update, which is here, because I noticed it while working on resource exhaustion issues. Would you prefer if I open separate GPII ticket and PR for this change?

**Comment by natarajaya** [comment]
Thanks for the reviews!

> Maybe region doesn't matter for Projects and DNS records?

@mrtyler I tested basic infra features (run locust, logging, monitoring) after switching region for my cluster with the code from this PR, everything seems to be working just fine. We probably don't want to have this discrepancy in regions for staging and production, but for dev clusters I think it is okay.

> There's also this gcp/README item that should be updated:

Updated.

> Since we're doing these changes, I would vote for dropping `additional_zones` config.

@stepanstipl But how can we get 4 worker nodes in this case? `initial_node_count` is the number of workers **per zone** – if we drop `additional_zones`, number of zones [defaults to 3](https://www.terraform.io/docs/providers/google/r/container_cluster.html#region), so we can have only 3, 6, 9... workers. The solution is obviously to use node pools instead, but I don't want to touch this here.

> In addition to this, `us-central1` is set in `gcp-stackdriver-export` and `gcp-secret-mgmt` modules, and should be changed there as well.

When left unchanged this does not break any functionality of the infra, but agree it makes sense to have GS storage in the same region as other resources. I'll update these modules too for consistency.

> Also Terraform state bucket is currently left at default

I would prefer to leave it as it is for now.

> changing a region seems unnecessarily complicated with current structure -> imho ii should be done at one place only

Agree. Added another "Too Hot For October" topic.

File: gcp/live/dev/infra/network/terraform.tfvars (hunk 2 above)
Comment (natarajaya) [reply gcp/live/dev/infra/network/terraform.tfvars]: Sounds good!

**Comment by stepanstipl** [comment]
Thanks for the updates @natarajaya, I think it simplifies things a bit -> well done 👍 

> @stepanstipl But how can we get 4 worker nodes in this case? `initial_node_count` is the number of workers **per zone** – if we drop `additional_zones`, number of zones [defaults to 3](https://www.terraform.io/docs/providers/google/r/container_cluster.html#region), so we can have only 3, 6, 9... workers. The solution is obviously to use node pools instead, but I don't want to touch this here.

Yes, 4 is tricky, but the question is why 4? We can live with 3 larger ones or 6 smaller ones, no? Being "stuck" with 4 is exactly the problem we end up with when using this "hardcoded" zones, specific for one region.

So unless we have some real reason to stick with 4, I would say let's not try to create and solve problems we don't have, and let's go with 3 or 6. (2 zones do not work very well anyway from the availability/redundancy point of view).

>> Also Terraform state bucket is currently left at default
>I would prefer to leave it as it is for now.

This doesn't sound right to me -> I decide to have a cluster in Europe and my state will be in US?

**Comment by natarajaya** [comment]
> Yes, 4 is tricky, but the question is why 4? We can live with 3 larger ones or 6 smaller ones, no?

@stepanstipl Of course we can use nodes with more CPU, it would look something like this:

**Development:**
Current: 4 * n1-standard-2 (4*$48.55=$194.2/month)
New: 3 * n1-highcpu-4 (3*$72.46=$217.38/month)

**Staging / Prod:**
Current: 4 * n1-highcpu-4 (4*$72.46=$289.84/month)
New: 3 * n1-highcpu-8 (3*$144.92=$434.76/month)

If team agrees on this migration (will cause cluster recreation), we can go forward and drop `additional_zones` from regional cluster configuration.

> This doesn't sound right to me -> I decide to have a cluster in Europe and my state will be in US?

TF state bucket is Multi-Regional with US as default location. Unfortunately, it is not easy to change the location, because it will cause bucket recreation and issues with immutable resources (KMS keys and keyrings, some other infra resources, more info in GPII-3241).

**Comment by stepanstipl** [comment]
> **Development:**
> Current: 4 * n1-standard-2 (4*$48.55=$194.2/month)
> New: 3 * n1-highcpu-4 (3*$72.46=$217.38/month)

> **Staging / Prod:**
> Current: 4 * n1-highcpu-4 (4*$72.46=$289.84/month)
> New: 3 * n1-highcpu-8 (3*$144.92=$434.76/month)

We already talked about bigger instances in the context of avoiding resource exhaustion, I would be up for this.

Note here - looks like additional_zones is now deprecated (in favor of node-locations, see https://cloud.google.com/sdk/gcloud/reference/container/clusters/create).

>> This doesn't sound right to me -> I decide to have a cluster in Europe and my state will be in US?

> TF state bucket is Multi-Regional with US as default location. Unfortunately, it is not easy to change the location, because it will cause bucket recreation and issues with immutable resources (KMS keys and keyrings, some other infra resources, more info in GPII-3241).

So what you're saying is that this change is non-trivial. I agree, but that is more of a reason not to leave this kind of technical debt behind. If we allow for region to be configured, our state should not be hardcoded to US.

Probably the strategy here would be to migrate existing state into new bucket? KMS rings we use are gobal, so if we migrate state nothing, but the bucket, needs to be recreated. I imagine this should not be such a big deal with some migration code:

As a weaker alternative - manage the appropriate multi-cluster location to match the region, ie. based on region choose correct one from `asia`, `eu` and `us`.

**Comment by natarajaya** [comment]
> We already talked about bigger instances in the context of avoiding resource exhaustion, I would be up for this in favor of consistency & simplicity.

>Note here - looks like additional_zones is now deprecated

@stepanstipl Alright then, I'll update node flavors for all environments and remove `additional_zones` variable from cluster configuration block.

> but that is more of a reason not to leave this kind of technical debt behind

I would not call this a technical debt, since it does not really affect any functionality of the infra (well, maybe having state bucket in EU can reduce total deployment time by couple of seconds when running Terraform from EU).

> Probably the strategy here would be to migrate existing state into new bucket?

I think so. We'll need the code that detects if current infra region is different from TF state bucket region, creates intermediate bucket ([you can't rename them](https://cloud.google.com/storage/docs/moving-buckets)), copies state for immutable resources into intermediate bucket, removes old bucket, creates new bucket with the same name but new location, copies resources again... We'll also need to handle this situation on common side, because TF state buckets are currently managed in [common code](https://github.com/gpii-ops/gpii-infra/blob/master/common/modules/gcp-project/main.tf#L326-L333). Seems like a lot of trouble.

I'll open a new ticket about making all infra resources (beside TF state bucket there are DNS resources and probably something else) consistent within same infra region.

**Comment by natarajaya** [comment]
Updated PR description with summary of changes and deployment plan.
I still need some more time to test that infra code can handle switching region with `TF_VAR_infra_region` gracefully, but all looks good so far.

**Comment by mrtyler** [comment]
LGTM though I may take another pass tomorrow (no need to wait for me when you're ready to merge). I agree with the cluster changes, and with handling the location of infra resources separately.

**Comment by mrtyler** [comment]
Oh, re: this deployment plan item:

> Disable cluster destruction protector.

We might try the method outlined in [this fortuitously-timed comment](https://github.com/gpii-ops/gpii-infra/pull/199#issuecomment-463017515).

**Comment by stepanstipl** [comment]
Cool, thanks @natarajaya, LGTM

> I'll open a new ticket about making all infra resources (beside TF state bucket there are DNS resources and probably something else) consistent within same infra region.

Sounds good to have this covered by another ticket.

**Comment by mrtyler** [comment]
I reviewed this again and LGTM. Good job working through the concerns and landing on a clean set of changes.

**Comment by natarajaya** [comment]
@mrtyler @stepanstipl Thanks for your reviews!

I tested this and discovered one small issues that is being addressed in (https://github.com/gpii-ops/exekube/pull/40), everything else works fine.

**Comment by stepanstipl** [comment]
thanks @natarajaya, looks good.

**Comment by amatas** [comment]
LGTM
